design-documents/csp.md

```diff
@@ -0,0 +1,208 @@
+        <policy id="font-src">
+            <values>
+                <value>fonts.mywebsite.com</value>
+            </values>
+        </policy>
+    </policies>
+</csp_whitelist>
+```
+ 
+#### Extension Developers
+Extension developers will be able to place _csp_whitelist.xml_ inside their modules just as theme developers. For more intricate
+cases they will be able to create a custom _PolicyCollectorInterface_ implementations to edit policies in the runtime.
+ 
+#### Per-page configuration
+It should be possible to configure CSP for a single page to allow access to certain resources/features
+for just that one document. To achieve this developers will be able to implement a special interface by the
+controller for the page they want custom CSP for.
+```php
+namespace Magento\Security\Model\CSP\PolicyCollector;
+
+interface CSPAwareActionInterface extends ActionInterface
+{
+    public function modifyCsp(PolicyInterface[] $appliedPolicies): PolicyInterface[];
+}
+```
+ 
+Thus controllers will be able to exclude/add policies to configured CSPs.
+
+## Reporting
+Browsers can report CSP violations in both modes. Magento will provide a default endpoint to receive these reports.
+Since there is no way to authenticate a genuine report and on a live store they can fill up quickly the number of
+reports stored in the database will be limited to 10000 deleting old ones when the limit is reached.
+ 
+Merchants will see new violations attention message in the admin panel.
+ 
+![alt text](img/csp_violations.png "CSP Violations")
+ 
+Extension developers will be able to redefine default report-uri via config and implement a
+custom _ViolationsRepositoryInterface_ to show them in the admin panel.
+ 
+Violations report should be accessible through web API. Both CSP related admin pages and the web API endpoints
+must require _Magento_Backend::system_ resource access.
+ 
+## Default CSP
+Magento uses some JavaScript libraries but usually has them provided with Magento itself without including
+them from 3rd party resources so for most cases having *-src directives with 'self' value will be enough. In some .phtmls
+we set event listener via attributes and we use eval in template.js so we'd have to also allow
+'unsafe-inline' and 'unsafe-eval'. Magento has integrations with multiple 3rd party services like vimeo, youtube,
+google analytics and various payment systems which we must whitelist additionally.
+That will be the default whitelist Magento provides out of the box packaged alongside related modules via _csp\_whitelist.xml_.
+ 
+## Future steps
+To allow merchants to use whitelisting with _nonce_ we have to get reed of event handlers via HTML attributes and
+style attributes in our templates. There is no way to disable `unsafe-eval` since we use it for UI components and
+some of the front-end libraries we employ need it (like jQuery).
+ 
+The work on refactoring templates to remove event handlers and _style_ attributes will be done after CSP is introduced
```

Comment:
> The work on refactoring templates to remove event handlers and _style_ attributes will be done after CSP is introduced
gradually

This feels like the wrong order to me. How can we except merchants to find the time to update their stores for CSP if we, Magento, can't find the time to do it first?

Reply:
A good example of this is the `@escapeNotVerified` security work done several years ago.

https://magento.stackexchange.com/questions/92963/magento-2-escapenotverified

We _still_ haven't completed this work in core, and it's been > 4 years.

Reply:
> We still haven't completed this work in core

Not to take away from your point, we just did it in the last few months. It will be delivered in the next patch releases.

To actually add to your point, it took the work of ALL core teams over 2 months of solid work to properly implement and test those changes. Totally unrealistic effort level for a typical merchant.

Reply:
even with inline styles/JS present in templates we can utilize origin whitelisting for remote resources